### PR TITLE
Fixed #32483 -- Fixed QuerySet.values()/values_list() on JSONField key transforms with booleans on SQLite.

### PR DESCRIPTION
**Pull request by felixxm**
ticket-32483

### CODE REVIEW

**Comment by felixxm** [comment]
I think it's a nice side-effect that we can get rid of a bit hacky workarounds for SQLite i.e. `JSONExact.process_lhs()` and `KeyTransformExact.process_lhs()`.

**Comment by felixxm** [comment]
buildbot, test on oracle.

**Comment by carltongibson** [comment]
Nice. I knew this would be about three lines once you'd finished it 😀 — I'll leave this here for others to play over the weekend 😉

**Comment by charettes** [comment]
I might be worth keeping a mention that the issue is still present for composite extracts though e.g.

```python
class DataModel(models.Model):
    data = models.JSONField()

obj = DataModel.objects.create(data={'composite': [True]})
composite = DataModel.objects.values_list('data__composite', flat=True).get()
assert composite[0] is True  # <- I think this will fail?
```

I suppose we'd need to recursively call your SQL logic against the extract to address this limitation.

**Comment by felixxm** [comment]
> It might be worth keeping a mention that the issue is still present for composite extracts though e.g.
> 
> ```python
> class DataModel(models.Model):
>     data = models.JSONField()
> 
> obj = DataModel.objects.create(data={'composite': [True]})
> composite = DataModel.objects.values_list('data__composite', flat=True).get()
> assert composite[0] is True  # <- I think this will fail?
> ```
> 
> I suppose we'd need to recursively call your SQL logic against the extract tree to address this limitation.

No, it doesn't work only for primitives, complex structures work fine. I will add more tests.

**Comment by felixxm** [comment]
@charettes I added more tests.

**Comment by charettes** [comment]
@felixxm awesome 🎉! So SQLite only has issues when dealing with boolean and null primitives, it's so strange.

**Comment by felixxm** [comment]
Yes it's :shrug:

**Comment by felixxm** [comment]
buildbot, test on oracle.

File: django/db/backends/sqlite3/operations.py [hunk 1]
```diff
@@ -21,6 +21,10 @@ class DatabaseOperations(BaseDatabaseOperations):
         'DateTimeField': 'TEXT',
     }
     explain_prefix = 'EXPLAIN QUERY PLAN'
+    # Datatypes that cannot be extracted properly, due to the way the
```
Comment (cjerdonek) [inline django/db/backends/sqlite3/operations.py:24]: "These are the types of data that ..." (PEP 8 wants comments to be complete sentences).

File: django/db/backends/sqlite3/operations.py [hunk 2]
```diff
@@ -21,6 +21,10 @@ class DatabaseOperations(BaseDatabaseOperations):
...
+    # JSON_EXTRACT() function is implemented on SQLite. Use JSON_TYPE()
+    # instead.
```
Comment (carltongibson) [inline django/db/backends/sqlite3/operations.py:26]:
```suggestion
    # List of datatypes to that cannot be extracted with JSON_EXTRACT() on
    # SQLite. Use JSON_TYPE() instead.
```

**Review by carltongibson — APPROVED** [review]
Looks good, thanks @felixxm ✨

**Comment by felixxm** [comment]
@carltongibson Thanks for checking :+1:

**Pull request merged**